### PR TITLE
updated parameter_sweep tool to allow for random sampling

### PR DESCRIPTION
**Pull request by jefalon**
## Fixes/Addresses:

None

## Summary/Motivation:

This PR add the ability to perform random sampling within the parameter sweep tool using uniform and/or normal distributions.

## Changes proposed in this PR:
- Adds Sample base class to keep the parameter sampling standardized
- Adds LinearSample, UniformSample, NormalSample classes, which are used to perform the respective sampling

### Legal Acknowledgement

By contributing to this software project, I agree to the following terms and conditions for my contribution:

1. I agree my contributions are submitted under the license terms described in the LICENSE.txt file at the top level of this directory.
2. I represent I am authorized to make the contributions and grant the license. If my employer has rights to intellectual property that includes these contributions, I represent that I have received permission to make contributions and grant the required license on behalf of that employer.

### CODE REVIEW

File: proteuslib/tools/parameter_sweep.py
Comment (bknueven) [inline proteuslib/tools/parameter_sweep.py:158]: Is there a use-case for this I'm missing? I took out explicit handling of indexed variables or parameters because it seemed like handing the parameter sweep tool an indexed pyomo object was likely to be an error. And if not, we need to handle this case in the `Sample` validation logic above.

File: proteuslib/tools/parameter_sweep.py
Comment (bknueven) [inline proteuslib/tools/parameter_sweep.py:228]: Here's where we could just check `isinstance(sweep_param[key], RandomSample)` or such if we used stronger typing.

File: proteuslib/tools/parameter_sweep.py
Comment (bknueven) [inline proteuslib/tools/parameter_sweep.py:404]: Similar to before, we should use `Bcast`; the non-zero ranks should know the size of `save_data` in advance.

**Review by bknueven — COMMENTED** [review]
Looks like a good start; I left lots of nits in the comments.

**Comment by eyoung55** [comment]
Still to do:
- [ ] Set consistent random seed
- [ ] Verify Gather preserves correct ordering
- [ ] Set type of buffer correctly
- [ ] Verify `num_samples` can be set correctly for all cases, including parsing tuple

File: proteuslib/tools/parameter_sweep.py
Comment (jefalon) [reply proteuslib/tools/parameter_sweep.py:50]: Agreed, this will make the typing more explicit. I've updated the classes

File: proteuslib/tools/parameter_sweep.py
Comment (jefalon) [reply proteuslib/tools/parameter_sweep.py]: updated this error message

File: proteuslib/tools/parameter_sweep.py
Comment (jefalon) [reply proteuslib/tools/parameter_sweep.py:117]: yeah good call, done.

File: proteuslib/tools/parameter_sweep.py
Comment (jefalon) [reply proteuslib/tools/parameter_sweep.py:166]: good point, this was just legacy code. I removed it

File: proteuslib/tools/parameter_sweep.py
Comment (jefalon) [reply proteuslib/tools/parameter_sweep.py:181]: Yeah I had to add the indexed logic to work with the 0D RO model. For some reason RO.A_comp is an indexed Var() and wasn't being updated. How do you want to go about treating this?

File: proteuslib/tools/parameter_sweep.py
Comment (jefalon) [reply proteuslib/tools/parameter_sweep.py]: I updated this logic to take advantage of the new FixedSample and RandomSample classes

File: proteuslib/tools/parameter_sweep.py
Comment (jefalon) [reply proteuslib/tools/parameter_sweep.py:117]: also, this is now handled in the _process_sweep_params() function which happen before this logic making this error message redundant. Should we keep it?

File: proteuslib/tools/parameter_sweep.py
Comment (bknueven) [reply proteuslib/tools/parameter_sweep.py:181]: In this case, the user would unroll the index themselves. E.g., RO.A_comp[0] would work, I believe.

File: proteuslib/tools/parameter_sweep.py
Comment (bknueven) [reply proteuslib/tools/parameter_sweep.py:117]: Yes, it doesn't hurt and makes the "case" analysis more explicit in this function. E.g., if we add another sampling method this function wouldn't silently accept it.

File: proteuslib/tools/parameter_sweep.py
Comment (bknueven) [reply proteuslib/tools/parameter_sweep.py:175]: I think the answer for the parameter_sweep tool and this other code is to break out the logic that does pre-processing to check for tuples/lists into its own function.

File: proteuslib/tools/parameter_sweep.py
Comment (adam-a-a) [reply proteuslib/tools/parameter_sweep.py:190]: FYI - A_comp is indexed by time and solvent component. I think this was originally done with dynamic modeling in mind (A_comp could change with time due to fouling although there are other ways we could account for membrane fouling over time). 

As for having the second index for solvent, this doesn't make much sense for RO since it supports only one solvent in the first place; I _think_ this was done for consistency with IDAES modeling and how things were normally indexed. On the other hand, for B_comp (salt permeability factor of membrane) this has indices for time and solute, but the RO model could technically take on more than one solute.

Comment (bknueven) [reply proteuslib/tools/parameter_sweep.py:190]: This is exactly why I'm nervous about setting indexed things -- it could be the source of silent bugs (setting indices you didn't mean to set), which might be very hard to detect. Further, given the structure of IDAES models the only "proper" use case I could imagine is fixing all the time indices of an IndexedVar or IndexedParam to the same value. Fixing all the non-time indices, like for `flow_mass_phase_comp`, would almost surely be an error.

File: proteuslib/tools/parameter_sweep.py
Comment (bknueven) [inline proteuslib/tools/parameter_sweep.py:262]: We could set `current_sample_type` to `FixedSample` as opposed to a string.

File: proteuslib/tools/parameter_sweep.py
Comment (bknueven) [reply proteuslib/tools/parameter_sweep.py:190]: I could be convinced that this could be a useful "typing aid" for the case of `A_comp`, e.g., and indexed var/param with exactly one index (in most use cases now). That said, we could check for that when we create the `Sample` object and replace the indexed component with the appropriate single var/param.

Comment (jefalon) [reply proteuslib/tools/parameter_sweep.py:191]: I tried to take the indexed A_comp and get our the actual Var() I wanted to change, but I couldn't get it to work. Perhaps I was doing it wrong

File: proteuslib/tools/parameter_sweep.py
Comment (jefalon) [reply proteuslib/tools/parameter_sweep.py]: removed

File: proteuslib/tools/parameter_sweep.py
Comment (jefalon) [reply proteuslib/tools/parameter_sweep.py]: added the check for single valued indexed sets

File: proteuslib/tools/parameter_sweep.py
Comment (jefalon) [reply proteuslib/tools/parameter_sweep.py]: added an Enum class

File: proteuslib/tools/parameter_sweep.py
Comment (bknueven) [inline proteuslib/tools/parameter_sweep.py:35]: Is there a reason to allow the pyomo_object to be None? If is it None, the current code will crash somewhere...

Comment (eyoung55) [reply proteuslib/tools/parameter_sweep.py:35]: I think I added this to simplify some of the unit tests, e.g., create a sample like `UniformSample(None, range_A[0], range_A[1])` to test only that the random pieces are working.  Admittedly, just passing in a dummy `pyomo_object` isn't much harder and would remove this extra check.

File: proteuslib/tools/parameter_sweep.py
Comment (bknueven) [inline proteuslib/tools/parameter_sweep.py:39]:
```suggestion
            if pyomo_object.is_indexed() and len(pyomo_object) == 1:
```

Comment (jefalon) [reply proteuslib/tools/parameter_sweep.py:39]: Whoops, looks like I left a bit of debugging in there. Thanks for cleaning it up!

Comment (bknueven) [reply proteuslib/tools/parameter_sweep.py:39]: It's tedious, but we should add a simple case for the unit tests. Note something as simple as 
```
m=pyo.ConcreteModel()
m.x = pyo.Var(['a'])
```
would trigger this logic when `m.x` is passed in as the `pyomo_object`

File: proteuslib/tools/parameter_sweep.py
Comment (adam-a-a) [inline proteuslib/tools/parameter_sweep.py:26]: Just out of curiosity, can we use `Enum` here or is `IntEnum` necessary here? If `IntEnum` _is_ necessary, is it because `auto()` is being used?

**Review by bknueven — APPROVED** [review]
The tests have some redundant code, but that is a minor issue. Additional documentation for these changes is part of #161.

File: proteuslib/tools/parameter_sweep.py
Comment (bknueven) [reply proteuslib/tools/parameter_sweep.py]: This is a good catch -- I looked it up and `IntEnum` will silently cast its members to `int` for comparison purposes. E.g.
```python
class FooType(Enum):
    BAR = 1

class BarType(IntEnum):
    FOO = 1

# False
FooType.BAR == 1

# True
BarType.FOO == 1
```

The `auto()` function is just for convenience.

**Review by bknueven — APPROVED** [review]
@jefalon I'm happy with this as-is. See if you're okay with the changes Ethan and I have made.

**Comment by jefalon** [comment]
It looks great to me! Thanks everyone for helping get this finished. As soon as we get that second approval, I think we are ready to merge.

**Review by eyoung55 — APPROVED** [review]
Looks good to me.

**Comment by jefalon** [comment]
@bknueven I don't think I have the right to merge. Do you?

**Pull request merged**
